bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/tasks/ExportFromModelServer.kt

```diff
@@ -0,0 +1,101 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.model.sync.bulk.gradle.tasks
+
+import kotlinx.coroutines.runBlocking
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Optional
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+import org.modelix.model.ModelFacade
+import org.modelix.model.api.IBranch
+import org.modelix.model.api.IProperty
+import org.modelix.model.api.PBranch
+import org.modelix.model.api.getRootNode
+import org.modelix.model.client2.IModelClientV2
+import org.modelix.model.client2.ModelClientV2
+import org.modelix.model.client2.ModelClientV2PlatformSpecificBuilder
+import org.modelix.model.client2.getReplicatedModel
+import org.modelix.model.data.NodeData
+import org.modelix.model.lazy.RepositoryId
+import org.modelix.model.mpsadapters.RepositoryLanguage
+import org.modelix.model.sync.bulk.ModelExporter
+import javax.inject.Inject
+
+abstract class ExportFromModelServer @Inject constructor(of: ObjectFactory) : DefaultTask() {
+
+    @Input
+    val url: Property<String> = of.property(String::class.java)
+
+    @Input
+    @Optional
+    val repositoryId: Property<String> = of.property(String::class.java)
+
+    @Input
+    @Optional
+    val branchName: Property<String> = of.property(String::class.java)
+
+    @Input
+    @Optional
+    val revision: Property<String> = of.property(String::class.java)
+
+    @OutputDirectory
+    val outputDir: DirectoryProperty = of.directoryProperty()
+
+    @TaskAction
+    fun export() {
+        val client = ModelClientV2PlatformSpecificBuilder().url(url.get()).build().apply { runBlocking { init() } }
+        val branch = if (revision.isPresent) {
+            getBranchByRevision(client)
+        } else {
+            getBranchByRepoIdAndBranch(client)
+        }
+
+        branch.runRead {
+            val root = branch.getRootNode()
+            logger.info("Got root node: {}", root)
+            val outputDir = outputDir.get().asFile
+            root.allChildren.forEach {
+                val nameRole = IProperty.fromName(RepositoryLanguage.NamePropertyUID)
+                val fileName = it.getPropertyValue(nameRole) ?: it.getPropertyValue(IProperty.fromName(NodeData.idPropertyKey))
+                    ?.substringBeforeLast(")", missingDelimiterValue = "null")
+                    ?.substringAfterLast("(", missingDelimiterValue = "null") // fallback if module name is not in properties
```

Comment:
That's pretty hard to read. Maybe a method or some intermediate variables explaining the pipeline steps could help understanding what's done here to the node to derive the filename

Reply:
It's an ugly hack that makes assumptions about the format of the ID. The export should not be MPS specific.